FIPS/fip-draft_foundational_fvm.md

```diff
@@ -0,0 +1,1010 @@
+  - [Syscalls](#syscalls-1)
+    - [Namespace: `actor`](#namespace-actor)
+    - [Namespace: `crypto`](#namespace-crypto)
+    - [Namespace: `gas`](#namespace-gas)
+    - [Namespace: `ipld`](#namespace-ipld)
+    - [Namespace: `message`](#namespace-message)
+    - [Namespace: `network`](#namespace-network)
+    - [Namespace: `rand`](#namespace-rand)
+    - [Namespace: `send`](#namespace-send)
+    - [Namespace: `self`](#namespace-self)
+    - [Namespace: `vm`](#namespace-vm)
+  - [Externs](#externs-1)
+  - [IPLD memory model](#ipld-memory-model)
+  - [Error conditions](#error-conditions)
+  - [Memory policy](#memory-policy)
+  - [Module caching](#module-caching)
+  - [Gas charging](#gas-charging)
+  - [Specific actor coupling](#specific-actor-coupling)
+- [Foundational FVM network upgrade](#foundational-fvm-network-upgrade)
+- [Design Rationale](#design-rationale)
+- [Backwards Compatibility](#backwards-compatibility)
+- [Test Cases](#test-cases)
+- [Security Considerations](#security-considerations)
+- [Incentive Considerations](#incentive-considerations)
+- [Product Considerations](#product-considerations)
+- [Implementation](#implementation)
+- [Copyright](#copyright)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Simple Summary
+
+User-programmability is the ability to deploy and run arbitrary user-provided
+code on some infrastructure, in this case a blockchain network.
+
+Today, the Filecoin blockchain lacks user-programmability. This FIP is part of a
+series that aims to introduce this capability.
+
+A change of this magnitude impacts many components. Concretely:
+
+1. The storage layer, to store and load user-provided code.
+2. The execution layer, to compile/interpret user-provided code, run state
+   transitions involving such code, and verify the results.
+3. The gas model, to secure the network and modulate chain capacity
+   demand/supply dynamics in the face of user-programmable code.
+4. The preexisting baseline protocol, to facilitate programmability against
+   system functionality.
+
+This FIP focuses on (1) and (2). The remaining points are out of scope of this
+FIP, and will be addressed by subsequent proposals.
+
+Moreover, this FIP proposes a network upgrade to introduce a foundational
+version of the FVM.
+
+## Abstract
+
+We specify the Filecoin Virtual Machine (FVM), a IPLD-ready WASM-based execution
+layer capable of running arbitrary user-provided code. The FVM replaces the
+existing execution layer.
+
+The logic of native actors (the equivalent of smart contracts in Filecoin) is
+deployed as WASM bytecode. Foreign actors targeting non-Filecoin runtimes such
+as the Ethereum Virtual Machine are also supported through runtime emulation.
+
+We provide a reference architecture for the FVM, and explain its various
+components: (a) Machine, (b) Call Manager, (c) Invocation Container, (d) Kernel,
+(e) Syscalls, and (f) Externs. We define the contract that actor code must
```

Comment:
I'm not sure if `Externs` have a place here. We've implemented the reference FVM that way, but a large part of that is the rust/go boundary.

Reply:
I think of the Externs as a logical boundary; it's a set of externally provided functions that the VM needs to operate. Indeed it is emphasised in the Go/Rust integration because in that case it becomes a _physical_ boundary due to the cross of technologies. I'll clarify this aspect in the doc, and we can continue to debate on the public FIP if that SGTY?